### ISSUE TITLE
test_secmem failure on solaris

### ISSUE THREAD

**Issue by mattcaswell** [openssl/openssl#5346]
Running latest master on x86 solaris:

````
$ V=1 make TESTS=test_secmem test
make depend && make _tests
( cd test; \
  mkdir -p test-runs; \
  SRCTOP=../. \
  BLDTOP=../. \
  RESULT_D=test-runs \
  PERL="/usr/local/bin/perl" \
  EXE_EXT= \
  OPENSSL_ENGINES=`cd .././engines; pwd` \
  OPENSSL_DEBUG_MEMORY=on \
    /usr/local/bin/perl .././test/run_tests.pl test_secmem )
../test/recipes/90-test_secmem.t .. 
1..1
    # Subtest: ../../test/secmemtest
    1..1
    # ERROR: (bool) 'CRYPTO_secure_malloc_init(4096, 32) == true' failed @ test/secmemtest.c:28
    # false
    not ok 1 - test_sec_mem
../../util/shlib_wrap.sh ../../test/secmemtest => 1
not ok 1 - running secmemtest

#   Failed test 'running secmemtest'
#   at /export/home/matt/dev/openssl/test/../util/perl/OpenSSL/Test/Simple.pm line 77.
# Looks like you failed 1 test of 1.
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 

Test Summary Report
-------------------
../test/recipes/90-test_secmem.t (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
Files=1, Tests=1,  0 wallclock secs ( 0.01 usr  0.01 sys +  0.07 cusr  0.04 csys =  0.13 CPU)
Result: FAIL
*** Error code 1
make: Fatal error: Command failed for target `_tests'
Current working directory /export/home/matt/dev/openssl
*** Error code 1
make: Fatal error: Command failed for target `tests'
````

**Comment by dot-asm** [comment]
I mentioned it in https://github.com/openssl/openssl/pull/5060#issuecomment-360786904. Basically suggestion is to amend the #if with defined(__sun) && defined(__SRV4) || defined(__hpux) || defined(__sgi) || defined(__osf__). Though one can argue if for example __hpux should be _HPUX_SOURCE...

**Comment by dot-asm** [comment]
Addressed in #5460.
